packages/realm-server/matrix-realm.ts

```diff
@@ -0,0 +1,136 @@
+import './e2ee';
+import { Deferred } from '@cardstack/runtime-common';
+import { MatrixSearchIndex } from './matrix-search-index';
+import { type MatrixClient, createClient } from 'matrix-js-sdk';
+
+interface Params {
+  matrixServerURL: string;
+  accessToken: string;
+  userId: string;
+  deviceId: string;
+  deferStartUp?: true;
+}
+
+export class MatrixRealm {
+  #startedUp = new Deferred<void>();
+  #deferStartup: boolean;
+  #client: MatrixClient;
+  #matrixServerURL: string;
+  #accessToken: string;
+  #deviceId: string;
+  #userId: string;
+  #searchIndex: MatrixSearchIndex;
+
+  constructor({
+    matrixServerURL,
+    accessToken,
+    userId,
+    deviceId,
+    deferStartUp,
+  }: Params) {
+    this.#matrixServerURL = matrixServerURL;
+    this.#accessToken = accessToken;
+    this.#deviceId = deviceId;
+    this.#userId = userId;
+    this.#client = createClient({ baseUrl: matrixServerURL });
+    this.#searchIndex = new MatrixSearchIndex(() => this.#client);
+
+    this.#deferStartup = deferStartUp ?? false;
+    if (!deferStartUp) {
+      this.#startedUp.fulfill((() => this.#startup())());
+    }
+  }
+
+  get ready(): Promise<void> {
+    return this.#startedUp.promise;
+  }
+
+  get #isLoggedIn() {
+    return this.#client.isLoggedIn();
+  }
+
+  // it's only necessary to call this when the realm is using a deferred startup
+  async start() {
+    if (this.#deferStartup) {
+      this.#startedUp.fulfill((() => this.#startup())());
+    }
+    await this.ready;
+  }
+
+  // primarily a test utility to await for message events to be indexed
+  async flushMessages() {
+    await this.#searchIndex.flushMessages();
+  }
+
+  // primarily a test utility to await for initial room/membership events to be
+  // indexed
+  async flushRooms() {
+    await this.#searchIndex.flushRooms();
+  }
+
+  shutdown() {
+    // note that it takes about 90 seconds to actually end the process after
```

Comment:
It takes over an hour for realm server tests to complete in CI due to this bug!!!! The issue has been languishing for years and is blocked on some major work. We should experiment to see if there is a hack we can use to get this to not hang.